src/worker.ts

```diff
@@ -132,32 +98,46 @@ export class Worker {
    * @returns true if the worker is healthy, false otherwise
    */
   async healthCheck() {
-    if (!this.#running) {
+    if (this.#process === null) {
       return false;
     }
-    const res = await fetch(`http://localhost:${this.port}`, {
-      headers: {
-        "X-Health-Check": "true",
-      },
-    });
-    return res.ok;
+    const startTime = Date.now();
+    while (Date.now() - startTime < 5000) {
+      try {
+        const res = await fetch(`http://localhost:${this.port}`, {
+          headers: {
+            "X-Health-Check": "true",
+          },
+        });
+        if (res.ok) {
+          return true;
+        }
+      } catch {
+        await new Promise((resolve) => setTimeout(resolve, 20));
+      }
+    }
+    return false;
   }
 
-  restart() {
-    console.log("[worker]", `${this.name}`, "restarting");
-    if (this.#running) {
-      this.#process?.kill("SIGINT");
-      this.#running = false;
+  shutdown() {
+    if (this.#process === null) {
+      return;
     }
-    return this.start();
+    console.log("[worker]", `${this.name}`, "shutting down");
+    this.#process.kill("SIGINT");
+    // todo(sbmsr): why does using a promise here break the console log flow?
+    this.#process.status.then((status) => {
+      console.log(
+        "[worker]",
+        `${this.name}::module(${this.module?.name}@${this.module?.version})`,
+        "stopped with",
+        status.code,
+      );
+      this.cleanupData();
+    });
```

Comment:
This may never be called? The process is killed first which means the signal likely is emitted before the signal listener ever gets registered for the process here. You could set up the status listener _before_ calling kill, but you're have to await it to ensure it was actually setup. I get that you're trying to make this synchronous though, and that would break your assumptions.